### ISSUE TITLE
Magic Rollback

### ISSUE THREAD

**Issue by Ekleog** [zhaofengli/colmena#38]
Hello,

I'm just now learning of colmena, and it looks great! I'm just opening this issue because I was originally looking for a flakeless alternative to deploy-rs, and having looked at it made me see its [magic rollback](https://github.com/serokell/deploy-rs#magic-rollback) feature.

How would you feel about implementing something like that for colmena?

I think the process would be to do something like:
1. upload the closure
2. run something akin to `nixos-rebuild test; (sleep 300 && rollback) &`
3. then open a new ssh connection (without ControlMaster, so with `ssh -S none`)
4. have the new ssh connection kill the `sleep 300 && rollback` command

I think that by using [grub's robustness mechanisms](https://www.gnu.org/software/grub/manual/legacy/Making-your-system-robust.html#Making-your-system-robust) it'd even be possible to have a similar mechanism working for when the system needs to be rebooted, which would be an awesome feature that AFAICT no other deployment mechanism has, but that'd certainly be a lot more work to implement.

**Comment by zhaofengli** [comment]
That's an interesting idea! I think an easy way to implement this is with a list of scripts (`deployment.checkScripts` maybe?) that are run on the target machine after the configuration is activated. If any of the scripts fail (e.g., internet connectivity check), the configuration is rolled back.

**Comment by Ekleog** [comment]
This would sound great! Though maybe even better if there were a way to get “default” check scripts so that everyone doesn't have to redevelop similar check scripts.

Actually that makes me think: it'd also require a `deployment.localCheckScripts` that would contain scripts ran on the deployer host, so that it could try to ssh into the being-deployed host, as forward-checks are not the same as backwards-checks if eg. the firewall is misconfigured in the new configuration.

**Comment by blaggacao** [comment]
I have to say I regularly have to switch off auto-rollback in `deploy-rs` because it's too indiscriminate.

So being able to define the checks oneself, is a great design.

Between that and a default, is the option of a library, still.

Defaults are comfortable, but intransparent.

**Comment by otavio** [comment]
I've been using colmena to deploy some DigitalOcean machines and while reworking the network config have lost connectivity few times, so having checks is very welcome.

**Comment by asymmetric** [comment]
@zhaofengli I noticed a comment mentioning an `auto-rollback.sh` script, but I can't find its source -- is this part of an implementation of this feature?

**Comment by NeverBehave** [comment]
> @zhaofengli I noticed [a comment](https://github.com/zhaofengli/colmena/blob/e034c15825c439131e4489de5a82cf8e5398fa61/src/nix/hive/assets.rs#L4) mentioning an `auto-rollback.sh` script, but I can't find its source -- is this part of an implementation of this feature?

Right now there is no such file. It is still a planning feature, but I am working on an alternative way to fit my need about rollback.

**Comment by NeverBehave** [comment]
Here is a quick idea how this would work, but still under planning. The idea about rollback has several ideas behind:

1. Assumes the last usable profile(target rollback) is current - 1
2. Flexible to implement user defined checks
3. System should be able to reboot and still do auto rollback (in case of applying kernel upgrade breaks network connections aka driver issue)

Here is an sample script:

```nix
let 
  timestamp = "/var/auto-rollback-timestamp";
  check-period = "*:0,30"; # every 30 mins
  wait-period = "30"; # allow 30 minutes to recover
in {
        deployment.keys."auto-rollback-timestamp" = {
                keyFile = ./blank; # An empty file
                destDir = "/var"; # Default: /run/keys
                permissions = "0755"; # Default: 0600
      
                uploadAt = "pre-activation";
        };

        system.activationScripts = {
            auto-rollback = {
                text = ''
                    # If auto-rollback-timestamp file exists and it is empty, write timestamp to file
                    if [ -f "${timestamp}" ]; then
                        if [ ! -s "${timestamp}" ]; then
                            date +%s > ${timestamp}
                        fi
                    fi
                '';
            };
        };

        systemd.timers.auto-rollback = {
            wantedBy = ["timers.target"];
            timerConfig = {
                OnCalendar = check-period;
                Unit = "auto-rollback.service";
            };
        };

        systemd.services.auto-rollback = {
            restartIfChanged = false;
            serviceConfig = {
            Type = "oneshot";
            ExecStart = pkgs.writeScript "auto-rollback" ''
                #!${pkgs.runtimeShell} -e

                timestamp="${timestamp}"

                function rollback_to_previous_profile {
                    rm $timestamp

                    # Find current generation
                    # Populate the array with file and folder names
                    while IFS= read -r file; do
                        files+=("$file")
                    done < <(ls -lt "/nix/var/nix/profiles" | ${pkgs.gawk}/bin/awk 'NR > 1 {print $NF}')
                    last_profile=''${files[2]}
                    echo "Switch back to Profile: $last_profile"
                    $last_profile/bin/switch-to-configuration switch
                }
                
                if [ ! -f "$timestamp" ]; then
                    echo "File does not exist: $timestamp, skipping rollback checks"
                    exit 0
                fi

                stored_timestamp=$(cat "$timestamp")
                number_regex='^[0-9]+$'

                if [[ $stored_timestamp =~ $number_regex ]]; then
                    echo "Valid timestamp: $stored_timestamp"
                else
                    echo "Invalid timestamp! Rolling back right now..."
                    rollback_to_previous_profile
                fi

                wait_period="${toString wait-period}"

                echo "checking timestamp $stored_timestamp"
                current_time=$(${pkgs.coreutils}/bin/date +%s)
                threshold=$((current_time - (wait_period * 60)))

                echo "Threshold $threshold"
                if [ "$stored_timestamp" -lt "$threshold" ]; then
                    echo "Current time has a difference of more than $wait_period minutes compared to the stored timestamp."
                    rollback_to_previous_profile
                else
                    echo "Current time is within $wait_period minutes of the stored timestamp."
                    exit 0
                fi
            '';
            };
        };
}
```

Here is an idea how the above scripts works: 

1. pre-activation: colmena will upload an empty file to given directory
2. activation script: if file exists and it is empty, write timestamp
3. The systemd services will repeatly check if the file exists and if corresponded timestamp has reach maximum wait time, delete the file and rollback
4. To stop rollback, delete `/var/auto-rollback-timestamp`

The good thing about this design it provide flexibly way for user to implement rollback: to stop it simply delete the file. Default I will do `colmena exec rm /var/auto-rollback-timestamp`, also user could run scripts that deletes the file. 

However the timer approach may introduce race condition and there is a constant timer in the system checking the file. This could be better addressed by have the systemd service trigger every time of deployment/onBootSec only. 

I am not entirely sure if this is a good approach at all, nor how this could be part of the colmena, or deploy-rs using similar strategy(which I should), but I leave it here for now and see it will help another needs a solution atm.

**Comment by JasmineLowen** [comment]
DISCLAIMER: I haven't tested the following idea yet. Use at your own risk.

Maybe we could also implement it as follows (just a rough idea, nothing specific yet). Before starting the main colmena process , `test` or `switch`, use `colmena exec` to track the `ssh` connection and if we are reasonably sure that the connection couldn't be established, reboot.

All in all this would be something like

- `colmena exec --on=abc "nohup sh -c script &"` where `script` is something like 
```
echo "This is just a weird identifier since I don't know how else to kill this process"
counter=0;
while :; do
  fin=$(ss -tnp | grep sshd)
  if [ $counter -ge 5 ] && [ -z $fin ]; then
    reboot
  else
    sleep 60;
    ((counter++))
  fi; 
done 
```
- `colmena apply test --on=abc`
- `colmena apply switch --on=abc` (optional)
- `colmena exec --on=abc ps aux | grep "weird identifier" | awk '{print $2}' | xargs -I{} kill {}`

As you can see some things, like stopping the process, are still a bit akward.

**Comment by eyJhb** [comment]
I think this could be very nice to have, but at the same time I think the scope described in this issue makes it an enormous task.
Essentially adding health checks would be a requirement for what is described here, which in itself is a pretty big task to do correctly.

Instead I like the approach that @NeverBehave have shown, as well as what both [Nixus](https://github.com/Infinisil/nixus) and [deploy-rs](https://github.com/serokell/deploy-rs) does it.
Basically, it's MVP, what do I need before I can correct any fault on the server? You need networking access + SSH access, and then you can remedy whatever issues that might be wrong.

I would instead purpose that we mirror that functionality as a start, and could tack any health checks on to it in the future.

What nixus does is basically the following ([source](https://github.com/infinisil/nixus/blob/master/scripts/switch)):

- Get current generation path from `/run/current-system`
- Activate the new system with `switch-to-configuration test`
- Wait for the deployer to make contact (confirms networking + ssh works)
- If contact is made, make it permanent: `switch-to-configuration boot` + set the system profile
- Write the result and release the lock
- If the test failed, the boot switch failed, or nobody made contact in time, roll back to the initial generation
- Wait for contact again to check the rollback worked
- If still nothing, reboot

The way the deployer confirms/indicates that it could make contact, is by continuously running `swiwch active <id>` (id is just incremental).
Which basically just writes to a FIFO file that the deployment script is waiting for, with a timeout in front.
